src/librustc_typeck/check/wfcheck.rs

```diff
@@ -521,7 +532,79 @@ fn check_fn_or_method<'a, 'fcx, 'gcx, 'tcx>(tcx: TyCtxt<'a, 'gcx, 'gcx>,
     // FIXME(#25759) return types should not be implied bounds
     implied_bounds.push(sig.output());
 
-    check_where_clauses(tcx, fcx, span, def_id);
+    check_where_clauses(tcx, fcx, span, def_id, Some(sig.output()));
+}
+
+fn check_existential_types<'a, 'fcx, 'gcx, 'tcx>(
+    tcx: TyCtxt<'a, 'gcx, 'gcx>,
+    fcx: &FnCtxt<'fcx, 'gcx, 'tcx>,
+    fn_def_id: DefId,
+    span: Span,
+    ty: Ty<'tcx>,
+) -> Vec<ty::Predicate<'tcx>> {
+    trace!("check_existential_types: {:?}, {:?}", ty, ty.sty);
+    let mut substituted_predicates = Vec::new();
+    ty.fold_with(&mut ty::fold::BottomUpFolder {
+        tcx: fcx.tcx,
+        fldop: |ty| {
+            if let ty::TyAnon(def_id, substs) = ty.sty {
+                trace!("check_existential_types: anon_ty, {:?}, {:?}", def_id, substs);
+                let anon_node_id = tcx.hir.as_local_node_id(def_id).unwrap();
+                if may_define_existential_type(tcx, fn_def_id, anon_node_id) {
+                    let generics = tcx.generics_of(def_id);
+                    trace!("check_existential_types may define. Generics: {:#?}", generics);
+                    for (subst, param) in substs.iter().zip(&generics.params) {
+                        if let ty::subst::UnpackedKind::Type(ty) = subst.unpack() {
+                            match ty.sty {
```

Comment:
This isn't quite good enough. We must also check that no type parameters are used *twice*. In particular, `fn foo<T>() -> Foo<T, T>` is wrong.